File: banchus-common/src/main/java/pe/nanamochi/banchus/util/Rijndael.java
package pe.nanamochi.banchus.util;

import org.bouncycastle.crypto.engines.RijndaelEngine;
import org.bouncycastle.crypto.modes.CBCBlockCipher;
import org.bouncycastle.crypto.paddings.PaddedBufferedBlockCipher;
import org.bouncycastle.crypto.paddings.ZeroBytePadding;
import org.bouncycastle.crypto.params.KeyParameter;
import org.bouncycastle.crypto.params.ParametersWithIV;

public final class Rijndael {

  private Rijndael() {
    throw new UnsupportedOperationException("Utility class");
  }

  public static byte[] encrypt(byte[] plaintext, byte[] key, byte[] iv) throws Exception {
    return process(true, plaintext, key, iv);
  }

  public static byte[] decrypt(byte[] ciphertext, byte[] key, byte[] iv) throws Exception {
    return process(false, ciphertext, key, iv);
  }

  private static byte[] process(boolean encrypt, byte[] data, byte[] key, byte[] iv)
      throws Exception {
    RijndaelEngine engine = new RijndaelEngine(256);

    PaddedBufferedBlockCipher cipher =
        new PaddedBufferedBlockCipher(CBCBlockCipher.newInstance(engine), new ZeroBytePadding());
    cipher.init(encrypt, new ParametersWithIV(new KeyParameter(key), iv));

    byte[] output = new byte[cipher.getOutputSize(data.length)];
    int len = cipher.processBytes(data, 0, data.length, output, 0);
    len += cipher.doFinal(output, len);

    byte[] result = new byte[len];
    System.arraycopy(output, 0, result, 0, len);
    return result;
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/core/util/Rijndael.kt
package pe.nanamochi.banchus.core.util

import org.bouncycastle.crypto.engines.RijndaelEngine
import org.bouncycastle.crypto.modes.CBCBlockCipher
import org.bouncycastle.crypto.paddings.PaddedBufferedBlockCipher
import org.bouncycastle.crypto.paddings.ZeroBytePadding
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.crypto.params.ParametersWithIV

object Rijndael {
    fun encrypt(data: ByteArray, key: ByteArray, iv: ByteArray): ByteArray =
        process(encrypt = true, data, key, iv)

    fun decrypt(data: ByteArray, key: ByteArray, iv: ByteArray): ByteArray =
        process(encrypt = false, data, key, iv)

    private fun process(
        encrypt: Boolean,
        data: ByteArray,
        key: ByteArray,
        iv: ByteArray,
    ): ByteArray {
        val engine = RijndaelEngine(256)
        val cipher =
            PaddedBufferedBlockCipher(CBCBlockCipher.newInstance(engine), ZeroBytePadding())

        cipher.init(encrypt, ParametersWithIV(KeyParameter(key), iv))

        val output = ByteArray(cipher.getOutputSize(data.size))
        val processedLen = cipher.processBytes(data, 0, data.size, output, 0)
        val finalLen = processedLen + cipher.doFinal(output, processedLen)

        return output.copyOfRange(0, finalLen)
    }
}
